File: ruby/atcoder/tokubeki10mon/09_ABC049C.rb
s = gets.strip

words = ['dream', 'dreamer', 'erase', 'eraser']

until s.empty? do
    reduced = false
    words.map do |w|
        if s[-w.size, w.size] == w
            s = s[0..-w.size - 1]
            reduced = true
            break
        end
    end

    unless reduced
        puts 'NO'
        exit 0
    end
end

puts 'YES'

File: go/atcoder/tokubeki10mon/09_abc049c.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	var s string
	fmt.Scan(&s)

	words := [4]string{"dream", "dreamer", "erase", "eraser"}
	ans := "YES"
	for s != "" {
		matched := false
		for _, word := range words {
			l := len(s)
			if strings.HasSuffix(s, word) {
				s = s[:l-len(word)]
				matched = true
				break
			}
		}

		if !matched {
			ans = "NO"
			break
		}
	}

	fmt.Println(ans)
}
